lib/src/clojure_lsp/kondo.clj

```diff
@@ -235,39 +237,82 @@
+        result)
+      (catch Exception e
+        (logger/error e "Error running clj-kondo on" err-hint)))))
+
+(defn run-kondo-on-paths! [paths db* {:keys [external?] :as normalization-config}]
+  (let [db @db*
+        internal? (not external?)
+        custom-lint-fn (when internal?
+                         #(custom-lint-project! (normalize % normalization-config)))]
+    (-> (config-for-paths paths db)
+        (assoc-in [:config :output :analysis :java-class-usages] internal?)
+        (shared/assoc-in-some [:custom-lint-fn] custom-lint-fn)
+        (run-kondo! (str "paths " (string/join ", " paths)))
+        (normalize normalization-config))))
 
 (defn run-kondo-on-paths-batch!
   "Run kondo on paths by partitioning the paths, with this we should call
   kondo more times but with fewer paths to analyze, improving memory."
-  [paths public-only? update-callback db*]
+  [paths normalization-config update-callback db*]
   (let [total (count paths)
-        batch-count (int (Math/ceil (float (/ total clj-kondo-analysis-batch-size))))]
-    (logger/info (str "Analyzing " total " paths with clj-kondo with batch size of " batch-count " ..."))
-    (if (<= total clj-kondo-analysis-batch-size)
-      (run-kondo-on-paths! paths public-only? db*)
-      (->> paths
-           (partition-all clj-kondo-analysis-batch-size)
-           (map-indexed (fn [index batch-paths]
-                          (logger/info "Analyzing" (str (inc index) "/" batch-count) "batch paths with clj-kondo...")
-                          (update-callback (inc index) batch-count)
-                          (run-kondo-on-paths! batch-paths public-only? db*)))
+        batches (->> paths
+                     (partition-all clj-kondo-analysis-batch-size)
+                     (map-indexed (fn [index batch-paths]
+                                    {:index (inc index)
+                                     :paths batch-paths})))
+        batch-count (count batches)]
+    (logger/info
+      (case batch-count
+        0 "No new paths to analyze"
+        1 (str "Analyzing " total " paths with clj-kondo")
+        (str "Analyzing " total " paths with clj-kondo in " batch-count " batches...")))
+    (case batch-count
+      0 {}
+      1 (run-kondo-on-paths! paths db* normalization-config)
+      (->> batches
+           (map (fn [{:keys [index paths]}]
+                  (logger/info "Analyzing" (str index "/" batch-count) "batch paths with clj-kondo...")
+                  (update-callback index batch-count)
+                  (run-kondo-on-paths! paths db* normalization-config)))
            (reduce shared/deep-merge)))))
 
 (defn run-kondo-on-reference-filenames! [filenames db*]
-  (catch-kondo-errors (str "files " (string/join ", " filenames))
-    (kondo/run! (kondo-for-reference-filenames filenames db*))))
+  (let [db @db*
+        normalization-config {:external? false
+                              :ensure-filenames filenames}
```

Comment:
There's a chance this fixes a bug, though I didn't confirm whether it was really a bug in the first place.

My hypothesis is that if before the initial file was changed a reference file had some findings, and that after the changes it had none, we were leaving the original findings in the db. With this `:ensure-filenames`, we should be clearing those findings.

Reply:
yeah, almost sure this fixes that behavior, I was about to debug why that happens, quite annoying, will check later if that fixed it indeed